sway/input/keyboard.c

```diff
@@ -335,66 +316,65 @@ static void handle_keyboard_key(struct wl_listener *listener, void *data) {
 	struct wlr_seat *wlr_seat = keyboard->seat_device->sway_seat->wlr_seat;
 	struct wlr_input_device *wlr_device =
 		keyboard->seat_device->input_device->wlr_device;
-	wlr_idle_notify_activity(keyboard->seat_device->sway_seat->input->server->idle, wlr_seat);
 	struct wlr_event_keyboard_key *event = data;
-	bool input_inhibited = keyboard->seat_device->sway_seat->exclusive_client != NULL;
+	bool input_inhibited =
+		keyboard->seat_device->sway_seat->exclusive_client != NULL;
 
-	xkb_keycode_t keycode = event->keycode + 8;
-	bool handled = false;
+	wlr_idle_notify_activity(
+			keyboard->seat_device->sway_seat->input->server->idle, wlr_seat);
 
-	// handle keycodes
-	handled = keyboard_execute_bindcode(keyboard, event, input_inhibited);
+	xkb_keycode_t keycode = event->keycode + 8;
 
-	// handle translated keysyms
-	if (!handled && event->state == WLR_KEY_RELEASED) {
-		handled = keyboard_execute_bindsym(keyboard,
-			keyboard->pressed_keysyms_translated,
-			keyboard->modifiers_translated,
-			event->state, input_inhibited);
-	}
+	// update translated keysyms
 	const xkb_keysym_t *translated_keysyms;
 	size_t translated_keysyms_len =
 		keyboard_keysyms_translated(keyboard, keycode, &translated_keysyms,
 			&keyboard->modifiers_translated);
 	pressed_keysyms_update(keyboard->pressed_keysyms_translated,
-		translated_keysyms, translated_keysyms_len, event->state);
-	if (!handled && event->state == WLR_KEY_PRESSED) {
+			translated_keysyms, translated_keysyms_len);
+
+	// update raw keysyms
+	const xkb_keysym_t *raw_keysyms;
+	size_t raw_keysyms_len =
+		keyboard_keysyms_raw(keyboard, keycode, &raw_keysyms,
+			&keyboard->modifiers_raw);
```

Comment:
You only generate keysyms from the last pressed keycode. You should use `wlr_keyboard.keycodes` instead of just `keycode`.